lib/Conversion/P4HIRToBMv2IR/P4HIRToBMv2IR.cpp

```diff
@@ -21,7 +32,211 @@ namespace P4::P4MLIR {
+        });
+
+        if (!transitionInserted) return failure();
+
+        op.walk([&](P4HIR::ParserAcceptOp acceptOp) {
+            ConversionPatternRewriter::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToEnd(&transitionBlock);
+            rewriter.create<BMv2IR::TransitionOp>(
+                acceptOp.getLoc(),
+                BMv2IR::TransitionKindAttr::get(context, BMv2IR::TransitionKind::Default), nullptr,
+                nullptr, nullptr);
+            eraseList.push_back(acceptOp.getOperation());
+        });
+        // TODO: p4c raises a warning "Explicit transition to reject not supported on this target"
+        //  for explicit transitions to reject
+        op.walk([&](P4HIR::ParserRejectOp rejectOp) {
+            ConversionPatternRewriter::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToEnd(&transitionBlock);
+            rewriter.create<BMv2IR::TransitionOp>(
+                rejectOp.getLoc(),
+                BMv2IR::TransitionKindAttr::get(context, BMv2IR::TransitionKind::Default), nullptr,
+                nullptr, nullptr);
+            eraseList.push_back(rejectOp.getOperation());
+        });
+
+        // Move the remaning parser ops to their region, they will be converted by the other
+        // patterns
+        newState.getParserOps().takeBody(op.getBody());
+        for (auto op : eraseList) rewriter.eraseOp(op);
+        rewriter.replaceOp(op, newState);
+
+        return success();
+    }
+
+ private:
+    static Operation *insertTransitionKey(Operation *op, ConversionPatternRewriter &rewriter,
+                                          Block *block, SmallVector<Operation*>& eraseList) {
+        auto loc = op->getLoc();
+        if (auto lookAheadOp = dyn_cast<P4CoreLib::PacketLookAheadOp>(op)) {
+            // TODO: not sure how to handle offsets
+            auto offset = rewriter.getI32IntegerAttr(0);
+            // TODO: can PacketLookAheadOp return something other than Bit?
+            auto bitTy = cast<P4HIR::BitsType>(lookAheadOp.getResult().getType());
+            auto width = rewriter.getI32IntegerAttr(bitTy.getWidth());
+            ConversionPatternRewriter::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToEnd(block);
+            eraseList.push_back(lookAheadOp);
+            return rewriter.create<BMv2IR::LookaheadOp>(loc, offset, width);
+        }
+        llvm_unreachable("Unsupported operand");
+    }
+
+    static BMv2IR::TransitionOp insertTransition(P4HIR::ParserSelectCaseOp caseOp,
+                                                 ConversionPatternRewriter &rewriter,
+                                                 Block *block) {
+        auto context = caseOp.getContext();
+        auto keysets = caseOp.getSelectKeys();
+        auto loc = caseOp.getLoc();
+
+        for (auto entry : keysets) {
+            return TypeSwitch<Operation *, BMv2IR::TransitionOp>(entry.getDefiningOp())
```

Comment:
you'd probably want to error out on masks and ranges here as default case?

Reply:
I've added a case and a test for masks, in BMv2 ranges are "broken down" into masks, and I think we have a pass for this as part of https://github.com/p4lang/p4mlir-incubator/pull/217